### DISCUSSION TITLE
Multi-Artist Feature

### DISCUSSION THREAD

**Discussion by SePHTaN** [ampache/ampache#2821] [Q&A]
Hello all,

as i belief, you all know that one of my own big wishes for ampache is the support of Multi-Artists.
Being here on a regular basis for quite some weeks now, i realised if you want it ... do it.
You all are busy with getting ampache modernized, php8.0 compliant, testible, raise code quality aso.
Hence there's not much room for you to achieve such a big change in a modest timeframe.

Here i want to drop myself in and contribute to ampache.
Now and then i will surely ask for some help cause i am unused working in a team and also with version control systems.
Also i want to push the kind how i work to a more professional style than using notepad++ as editor. 

**Ready or Not:**
In order to be able to start i tried to prepare myself and my workspace as well as possible.
If i am missing something or you recommend something different, then i will gladly change my setup.
Installed : Visual Studio Code 1.54.2
Installed : Git for Windows v2.30.2
Configured : Git and Github for work with VS Code (not sure for merges to origin)
Forked : ampache
Cloned : develop to my workspace (not sure if this is the best starting branch for feature extension)
Undone : start new branch "multi-artist"
Missing : what to best use in VS Code for php8.0 compliance help; other recommended plugins ?
Missing : VS Code setting ? vim:set softtabstop=4 shiftwidth=4 expandtab: ; help appriciated

Enough for introduction, next post i will roughly outline how i want to achieve multi-artist.
This will need to change ampache at a lot of places in a deeply manner, and there are for sure more than one possibility to achieve the goal.
So i will put all approaches here for discussion.

One thing i almost forgot, to introduce myself:

I am Stephan, 
heading for double 5 anniversary,
father of 3 children (1 daughter, 2 boys) 
teamleader in the gardening department of a (big) hardware store

**Comment by SePHTaN** [comment]
Just decided to take a smaller "project" to start.
#1245 #1365  Some people requested to include Album Art, Fix for encoding issue, m3u playlist into the zip-archive when downloading.
Goin at first with the minimal solution to just include the cover.
Its a relatively small task giving me some training for the workflow.

**Comment by wagnered** [comment]
@SePHTaN I believe  #1365 is no longer valid now. Or not totally valid.   The cover/artist art can be written to the file now and should be downloaded with the music file.  Adding a playlist seems like a good idea though.

As for #1245 I'm not sure if this request is valid anymore.   Today, I think most systems use UTF8 encoding which would display ç correctly. From my research, Zip utilities depend on the end-user system to determine encoding, not the content of the zip file.

Can't help you with VS. I have a Linux system as I think most contributors do. You might try a "live" Linux operating system.  That is a sysyem that boots off a CD or flash drive and runs in memory.  This way you could try out a Linux system and preserve your Windows system.  or if you have enough disk space, Linux could be installed creating a dual-boot system.

**Reply by SePHTaN** [reply]
All OK with VS got editor settings as wanted, also code inspection by plugin intelephense should be configured correct.

#1245 is about Encoding issue, i had the same problem with an album from Shakira it wasn't downloadable.
src/Module/UtilZiHandler.php wasn't able to read one file so the zipping of that album stops with that file.
Honestly i don't have a clue how to circumvent this. Best would be to not have files with weird encodings at all.
I asked the maintainer of eac to offer an option like musicbrainz picard to convert all non ascii-characters for foldernames
cause musicbrainz only converts song names when writing, but he rejected.

Including cover for album, full artist and playlist i have correctly implemented yet.
I have never understand where the usescenario is for the playlist file of an full album is, every player i used up till now had the possibility to just add content of zip/folder aso. to the play queue.
For a playlist its something different.
When i am right has ampache a playlist export functionality, maybee i can use the $media_files list zu zip as input the the playlist exporter and include the output into the zip.

What's the difference of:
AmpConfig and ConfigContainerInterface seams to me as at least partly duplicates
AmpConfig refers to ampache.cfg.php and ConfigContainerInterface more to Database settings ?
Playlist would get trickier

**Reply by usox** [reply]
> AmpConfig and ConfigContainerInterface seams to me as at least partly duplicates
> AmpConfig refers to ampache.cfg.php and ConfigContainerInterface more to Database settings ?

AmpConfig ist the "legacy" way to read config values. As those static calls like `AmpConfig::get('something')` are not testable, "ConfigContainer" was added as successor for AmpConfig.

**Reply by SePHTaN** [reply]
> AmpConfig ist the "legacy" way to read config values. As those static calls like `AmpConfig::get('something')` are not testable, "ConfigContainer" was added as successor for AmpConfig.

Good to know, so i am going to use ConfigContainer and exchange occurances of AmpConfig, ConfigurationKeyEnum needs to be extended for every config parameter of course.
Some aspects of OOP are still tough for me, currently i am struggling a little bit with using LegacyLogger in VAInfo.
In `public static function parse_pattern` vscode moans about : 
  `$this->logger->debug() : Cannot use '$this' in non-object context` makes sense function is static
  `LegacyLogger::debug() : Non static method 'debug' should not be called statically` also makes sense method isn't static.
What would be the right way to invoke LegacyLogger here, need to do some reading tomorrow :)

**Comment by usox** [comment]
Hi Stehpan :)
Nice introduction!

> Installed : Visual Studio Code 1.54.2

Maybe consider taking a look at phpstorm. Working on a huge project like ampache could be hard with vscode, phpstorm is way more sophisticated (but yes, it's not free),

> teamleader in the gardening department of a (big) hardware store

hehe, now I know who to ask out regarding my chili and tomato plants :laughing:

**Reply by SePHTaN** [reply]
About 2-3 weeks, then the first tomatos will be available :)
I harvested about 30kg last year

**Comment by mitchray** [comment]
Multi-artist is in now so closing :)
